driver-core/src/main/com/mongodb/internal/authentication/AzureCredentialHelper.java

```diff
@@ -31,25 +35,44 @@
-                + "/metadata/identity/oauth2/token?api-version=2018-02-01&resource=https://vault.azure.net";
-
-        Map<String, String> headers = new HashMap<>();
-        headers.put("Metadata", "true");
-        headers.put("Accept", "application/json");
-
-        String response = getHttpContents("GET", endpoint, headers);
-        try {
-            BsonDocument responseDocument = BsonDocument.parse(response);
-            if (responseDocument.containsKey("access_token")) {
-                return new BsonDocument("accessToken", responseDocument.get("access_token"));
-            } else {
-                throw new MongoClientException("The access_token is missing from Azure IMDS metadata response.");
+    private static final String ACCESS_TOKEN_FIELD = "access_token";
+    private static final String EXPIRES_IN_FIELD = "expires_in";
+
+    private static ExpirableValue<String> cachedAccessToken = ExpirableValue.expired();
+
+    public static synchronized BsonDocument obtainFromEnvironment() {
+        String accessToken;
+        Optional<String> cachedValue = cachedAccessToken.getValue();
+        if (cachedValue.isPresent()) {
+            accessToken = cachedValue.get();
+        } else {
+            String endpoint = "http://" + "169.254.169.254:80"
+                    + "/metadata/identity/oauth2/token?api-version=2018-02-01&resource=https://vault.azure.net";
+
+            Map<String, String> headers = new HashMap<>();
+            headers.put("Metadata", "true");
+            headers.put("Accept", "application/json");
+
+            BsonDocument responseDocument;
+            try {
+                responseDocument = BsonDocument.parse(getHttpContents("GET", endpoint, headers));
+            } catch (JsonParseException e) {
+                throw new MongoClientException("Exception parsing JSON from Azure IMDS metadata response.", e);
+            }
+
+            if (!responseDocument.isString(ACCESS_TOKEN_FIELD)) {
+                throw new MongoClientException(String.format(
+                        "The %s field from Azure IMDS metadata response is missing or is not a string", ACCESS_TOKEN_FIELD));
+            }
+            if (!responseDocument.isString(EXPIRES_IN_FIELD)) {
+                throw new MongoClientException(String.format(
+                        "The %s field from Azure IMDS metadata response is missing or is not a string", EXPIRES_IN_FIELD));
             }
-        } catch (JsonParseException e) {
-            throw new MongoClientException("Exception parsing JSON from Azure IMDS metadata response.", e);
-        }
+            accessToken = responseDocument.getString(ACCESS_TOKEN_FIELD).getValue();
+            int expiresInSeconds = Integer.parseInt(responseDocument.getString(EXPIRES_IN_FIELD).getValue());
+            cachedAccessToken = ExpirableValue.unexpired(accessToken, Duration.ofSeconds(expiresInSeconds).minus(Duration.ofMinutes(1)));
```

Comment:
Create the new cached access token with the expiration as the duration from the HTTP response minus one minute